graph-builder/src/registry.rs

```diff
@@ -37,24 +46,163 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
+fn trim_protocol(src: &str) -> &str {
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+fn get_credentials_or_empty_reader(credentials_path: &Option<PathBuf>) -> Box<dyn std::io::Read> {
```

Comment:
If you just inline this function, you can avoid the heap allocation. I don't think this helps much with readability.

Reply:
Using `Box<dyn std::io::Read>` as the return type is the only way I've found to tell the compiler that all cases yield a reader. I've double-checked inlining this yesterday and couldn't make it work. Do you want me to triple-check?

Reply:
I managed to inline the functionality yet the `Box` is still required. I suggest we leave it in and revisit on subsequent compiler releases.